File: python-selenium-pytest-bdd/src/saucedemo_uitests/pages/login_page.py
from selenium.webdriver.common.by import By

from saucedemo_uitests.pages.base_page import BasePage, Locator

_USERNAME_INPUT: Locator = (By.CSS_SELECTOR, "[data-test='username']")
_PASSWORD_INPUT: Locator = (By.CSS_SELECTOR, "[data-test='password']")
_LOGIN_BUTTON: Locator = (By.CSS_SELECTOR, "[data-test='login-button']")
_ERROR_BANNER: Locator = (By.CSS_SELECTOR, "[data-test='error']")


class LoginPage(BasePage):
    def submit_login(self, username: str, password: str) -> None:
        """Submits the login form without asserting the outcome - the caller decides whether success or
        a specific error is expected."""
        self.type_text(_USERNAME_INPUT, username)
        self.type_text(_PASSWORD_INPUT, password)
        self.click(_LOGIN_BUTTON)

    def get_error_message(self) -> str:
        return self.text_of(_ERROR_BANNER)

    def has_error(self) -> bool:
        return self.wait_and_check_visible(_ERROR_BANNER)

File: java-selenium-junit5-cucumber/src/test/java/com/saucedemo/uitests/pages/LoginPage.java
package com.saucedemo.uitests.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

public final class LoginPage extends BasePage {

    private static final By USERNAME_INPUT = By.cssSelector("[data-test='username']");
    private static final By PASSWORD_INPUT = By.cssSelector("[data-test='password']");
    private static final By LOGIN_BUTTON = By.cssSelector("[data-test='login-button']");
    private static final By ERROR_BANNER = By.cssSelector("[data-test='error']");

    public LoginPage(WebDriver driver) {
        super(driver);
    }

    /**
     * Submits the login form without asserting the outcome - the caller decides whether success or a
     * specific error is expected. Note this returns as soon as the click is dispatched; it does not wait
     * for the inventory page to finish rendering (see {@link InventoryPage#clearCart()}).
     */
    public void submitLogin(String username, String password) {
        typeText(USERNAME_INPUT, username);
        typeText(PASSWORD_INPUT, password);
        click(LOGIN_BUTTON);
    }

    public String errorMessage() {
        return textOf(ERROR_BANNER);
    }

    public boolean hasError() {
        return waitAndCheckVisible(ERROR_BANNER);
    }
}
